### PR TITLE
Reduction example

### PR DESCRIPTION
**Pull request by magnatelee**

### CODE REVIEW

File: examples/reduction/examples/bincount.py
Comment (jjwilke) [inline examples/reduction/examples/bincount.py:36]: Why is this here? Wouldn't the print force execution? Why do we have to consume the result?

File: examples/reduction/examples/matmul.py
Comment (jjwilke) [inline examples/reduction/examples/matmul.py:72]: maybe say "first exterior"

File: examples/reduction/examples/matmul.py
Comment (jjwilke) [inline examples/reduction/examples/matmul.py:75]: may say "second exterior"

File: examples/reduction/examples/matmul.py
Comment (jjwilke) [inline examples/reduction/examples/matmul.py:61]: Same question here. Why is this needed? what makes it "bogus"?

File: examples/reduction/examples/matmul.py
Comment (jjwilke) [inline examples/reduction/examples/matmul.py:36]: this way of creating a temp array and then assigning through a slice via a cunumeric array was a bit odd to me - and might look a bit funny in a tutorial.

If we're going to depend on cunumeric, I would just create it as a cunumeric array and add a helper function `get_cunumeric_store`?

File: examples/reduction/reduction/reduction.py
Comment (jjwilke) [inline examples/reduction/reduction/reduction.py:99]: same comment here. It seems more intuitive to me to create a cunumeric zeros and then get the store from it.

File: examples/reduction/examples/matmul.py
Comment (jjwilke) [inline examples/reduction/examples/matmul.py:24]: Personal preference would be to only have a `print_store` function and that be the only place we need a `to_cunumeric_array` function. It seems more intuitive to me to create cunumeric arrays and fetch stores from them rather than create a store and then convert it to a cunumeric array.

File: examples/reduction/examples/histogram.py
Comment (jjwilke) [inline examples/reduction/examples/histogram.py:33]: Same comment here. It was a bit weird to me to create a cunumeric array from the store, then assign to a slice of it from a numpy array. Might be more intuitive to create a cunumeric random array and then have a `get_cunumeric_store` helper function?

File: examples/reduction/reduction/reduction.py
Comment (jjwilke) [inline examples/reduction/reduction/reduction.py:42]: Should this just be a part of the Store? Shouldn't we just have `Store` define its own `__legate_data_interface__` ?

File: examples/reduction/examples/matmul.py
Comment (csadorf) [reply examples/reduction/examples/matmul.py:24]: In general, it would probably be better to to implement the `__str__` function of the store (wrapper) so that `print(store)` would yield the desired result.

File: examples/reduction/reduction/reduction.py
Comment (csadorf) [reply examples/reduction/reduction/reduction.py:42]: Yes, that seems like a better approach. I assume that the implementation here serves both as a stop-gap measure and also guides and motivates the implementation of such needed features.

**Comment by magnatelee** [comment]
@jjwilke I addressed your comments in the latest commit, so please take another look. (I had to wait until the core type system is merged to do this properly.) Note that I simply replaced `to_cunumeric_array` with `np.asarray`, as extracting a store from a cunumeric ndarray is way more verbose than wrapping a store with an ndarray and calling a cunumeric method on it. Let me know if you still think there's a better way.

File: examples/reduction/src/histogram.cc
Comment (jjwilke) [inline examples/reduction/src/histogram.cc:47]: may wish to have a clarifying comment:

// Use naive algorithm that loops all bin lower bounds to find match

File: src/core/mapping/mapping.cc
Comment (jjwilke) [inline src/core/mapping/mapping.cc:214]: Why is this change needed? This seems like it could have major implications. Is it required in this PR?

**Review by jjwilke — APPROVED** [review]
Overall LGTM. Couple minor questions/nits.

File: examples/reduction/src/histogram.cc
Comment (magnatelee) [reply examples/reduction/src/histogram.cc:48]: https://github.com/nv-legate/legate.core/pull/660/commits/7b552355ba524569ebcf84760e825e1fde955f6c

File: src/core/mapping/mapping.cc
Comment (magnatelee) [reply src/core/mapping/mapping.cc:214]: This and other changes for `tree_reduce` have been factored out into #699, so this diff will disappear once this PR is rebased on it. FYI, this diff actually fixes a performance problem for unbound stores; Legion's layout constraint check is currently too strict for single-field instances and it ends up refusing to reuse instances from unbound stores for downstream tasks even when they are reusable.

**Pull request merged**
